src/NuGet.Core/NuGet.PackageManagement/NuGetPackageManager.cs

```diff
@@ -478,46 +482,77 @@ private void InitializePackagesFolderInfo(string packagesFolderPath, bool exclud
                 throw new ArgumentNullException(nameof(secondarySources));
             }
 
-            if (nuGetProject is INuGetIntegratedProject)
+            var maxTasks = Environment.ProcessorCount;
+            var tasks = new List<Task<IEnumerable<NuGetProjectAction>>>(maxTasks);
+            var nugetActions = new List<NuGetProjectAction>();
+            
+            var buildIntegratedProjects = nuGetProjects.OfType<BuildIntegratedNuGetProject>().ToList();
+            var nonBuildIntegratedProjects = nuGetProjects.Except(buildIntegratedProjects).ToList();
+
+            // add tasks for all build integrated projects
+            foreach (var project in buildIntegratedProjects)
             {
+                // if tasks count reachs max then wait until an existing task is completed
+                if (tasks.Count == maxTasks)
```

Comment:
I believe in the onbuild restore uses just 4 threads. The problem here is that VS already uses a lot of memory, and there is no switch to disable parallel in the VSIX.  So if your thread count is too high there's really no workaround apart from restoring from the command line.